"dotnet package update" modifies wrong project file (csproj)

@Joe-K-Sewell commented on [Thu, 25 Sep 2025 22:36:16 GMT](https://github.com/dotnet/sdk/issues/51012)
### Describe the bug

With the CLI version `10.0.100-rc.1.25451.107`, running `dotnet package update` on one project intending to update PackageReferences in that project will instead add PackageReferences to a different project file that the first project references.

### To Reproduce

[Repository that reproduces the issue](https://github.com/Joe-K-Sewell/bug-repros/tree/main/dotnet-package-update).

1. Create the following:
    * A solution (`dotnet new sln`)
    * Class Library project Example.Core (`dotnet new classlib`)
    * xUnit project Example.Test (`dotnet new xunit`)
2. Reference the class library project from the xUnit test project.
    * From the Example.Test directory: `dotnet reference add ..\Example.Core\Example.Core.csproj --project .\Example.Tests.csproj`
3. Add the projects to the solution.
    * From the solution directory: `dotnet sln add .\Example.Core\Example.Core.csproj .\Example.Tests\Example.Tests.csproj`
4. Update the package references in the test project to their latest versions:
    * From the solution directory: `dotnet package update --project .\Example.Tests\Example.Tests.csproj`
    * **OR** from the Example.Test directory: `dotnet package update`

**Expected behavior**: changes are made to the `<PackageReference>` elements in `Example.Tests.csproj`.

**Actual behavior**: new `<PackageReference>` elements are added to `Example.Core.csproj`. [Example commit showing this behavior](https://github.com/Joe-K-Sewell/bug-repros/commit/e8d2cea24bb996ce8602cee4e909dfbb689cb6f9). This causes builds of the solution to fail:

```
PS C:\code\_real-name\bug-repros\dotnet-package-update> dotnet build
    C:\code\_real-name\bug-repros\dotnet-package-update\Example.Tests\Example.Tests.csproj : error NU1605:
      Warning As Error: Detected package downgrade: xunit.v3 from 3.0.1 to 3.0.0. Reference the package directly from the project to select a different version.
       Example.Tests -> Example.Core -> xunit.v3 (>= 3.0.1)
       Example.Tests -> xunit.v3 (>= 3.0.0)
```

### Exceptions (if any)

When trying to run the errant command with `-v diagnostic` for more info:

```
System.InvalidCastException: Specified cast is not valid.
   at System.CommandLine.Binding.ArgumentConverter.ConvertObject(ArgumentResult argumentResult, Type type, Object value)
   at System.CommandLine.Binding.ArgumentConverter.ConvertIfNeeded(ArgumentConversionResult conversionResult, Type toType)
   at System.CommandLine.Parsing.OptionResult.GetValueOrDefault[T]()
   at System.CommandLine.Parsing.SymbolResult.GetValue[T](String name)
   at Microsoft.DotNet.Cli.Extensions.ParseResultExtensions.SafelyGetValueForOption[T](ParseResult parseResult, String name)
   at Microsoft.DotNet.Cli.Telemetry.TelemetryFilter.LogVerbosityForAllTopLevelCommand(ICollection`1 result, ParseResult parseResult, String topLevelCommandName, Dictionary`2 measurements)
   at Microsoft.DotNet.Cli.Telemetry.TelemetryFilter.Filter(Object objectToFilter)
   at Microsoft.DotNet.Cli.Utils.TelemetryEventEntry.SendFiltered(Object o)
   at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, TimeSpan startupTime)
   at Microsoft.DotNet.Cli.Program.Main(String[] args)
```

### Further technical details

<details><summary>details of dotnet --info</summary>
<p>

```
.NET SDK:
 Version:           10.0.100-rc.1.25451.107
 Commit:            2db1f5ee2b
 Workload version:  10.0.100-manifests.0e2d47c4
 MSBuild version:   17.15.0-preview-25451-107+2db1f5ee2

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.26100
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\10.0.100-rc.1.25451.107\

.NET workloads installed:
There are no installed workloads to display.
Configured to use workload sets when installing new manifests.
No workload sets are installed. Run "dotnet workload restore" to install a workload set.

Host:
  Version:      10.0.0-rc.1.25451.107
  Architecture: x64
  Commit:       2db1f5ee2b

.NET SDKs installed:
  7.0.203 [C:\Program Files\dotnet\sdk]
  10.0.100-rc.1.25451.107 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 10.0.0-rc.1.25451.107 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 10.0.0-rc.1.25451.107 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.16 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 10.0.0-rc.1.25451.107 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  None

Environment variables:
  DOTNET_CLI_TELEMETRY_OPTOUT              [1]

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

</p>
</details>

- Reproduced with Powershell terminal in Visual Studio Code 1.104.2
    ```
    Version: 1.104.2 (user setup)
    Commit: e3a5acfb517a443235981655413d566533107e92
    Date: 2025-09-24T11:21:37.073Z (1 day ago)
    Electron: 37.3.1
    ElectronBuildId: 12404162
    Chromium: 138.0.7204.235
    Node.js: 22.18.0
    V8: 13.8.258.31-electron.0
    OS: Windows_NT x64 10.0.26100
    ```
- C# Dev Kit version: 1.50.51

--- a/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Package/Update/PackageUpdateCommandRunner.cs
+++ b/src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/Package/Update/PackageUpdateCommandRunner.cs
@@ -81,9 +81,7 @@ internal static class PackageUpdateCommandRunner
             return ExitCodes.Error;
         }
 
-        PackageSpec projectSpec = dgSpec.Projects.Count == 1
-            ? dgSpec.Projects[0]
-            : dgSpec.GetProjectSpec(dgSpec.Restore[0]);
+        PackageSpec projectSpec = dgSpec.GetProjectSpec(dgSpec.Restore[0]);
 
         // 2. Find suitable version of package(s) to update
         // Source provider will be needed to find the package version and to restore, so create it here.
@@ -141,7 +139,7 @@ internal static class PackageUpdateCommandRunner
             }
         }
 
-        var projectName = Path.GetFileNameWithoutExtension(dgSpec.Projects[0].FilePath);
+        var projectName = Path.GetFileNameWithoutExtension(projectSpec.FilePath);
         logger.LogInformation($"  {projectName}:");
 
         foreach (var packageResult in packagesToUpdateResult)
@@ -154,7 +152,7 @@ internal static class PackageUpdateCommandRunner
 
         // 3. Preview restore to validate changes
         logger.LogDebug(Strings.PackageUpdate_PreviewRestore);
-        var updatedDgSpec = GetUpdatedDependencyGraphSpec(dgSpec, packagesToUpdateResult);
+        var updatedDgSpec = GetUpdatedDependencyGraphSpec(projectSpec, dgSpec, packagesToUpdateResult);
         var restorePreviewResult = await packageUpdateIO.PreviewUpdatePackageReferenceAsync(updatedDgSpec, logger, cancellationToken);
 
         if (!restorePreviewResult.Success)
@@ -165,7 +163,7 @@ internal static class PackageUpdateCommandRunner
 
         // 4. Update MSBuild files
 
-        var updatedPackageSpec = updatedDgSpec.Projects[0];
+        var updatedPackageSpec = updatedDgSpec.GetProjectSpec(projectSpec.FilePath);
         int updatedCount = 0;
 
         foreach (var packageResult in packagesToUpdateResult)
@@ -555,9 +553,9 @@ internal static class PackageUpdateCommandRunner
         return result;
     }
 
-    private static DependencyGraphSpec GetUpdatedDependencyGraphSpec(DependencyGraphSpec currentDgSpec, List<PackageUpdateResult> packagesToUpdate)
+    private static DependencyGraphSpec GetUpdatedDependencyGraphSpec(PackageSpec projectSpec, DependencyGraphSpec currentDgSpec, List<PackageUpdateResult> packagesToUpdate)
     {
-        var updatedPackageSpec = currentDgSpec.Projects[0].Clone();
+        var updatedPackageSpec = projectSpec.Clone();
 
         foreach (var packageResult in packagesToUpdate)
         {

--- a/test/NuGet.Core.Tests/NuGet.CommandLine.Xplat.Tests/Commands/Package/Update/PackageUpdateCommandRunnerTests/MultiProjectTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.CommandLine.Xplat.Tests/Commands/Package/Update/PackageUpdateCommandRunnerTests/MultiProjectTests.cs
@@ -30,15 +30,15 @@ public class MultiProjectTests
         var solutionDirectory = RuntimeEnvironmentHelper.IsWindows
             ? @"S:\path\to\repo\src"
             : @"/path/to/repo/src";
-        var project1Path = Path.Combine(solutionDirectory, "Project1", "Project1.csproj");
+        var project1Path = Path.Combine(solutionDirectory, "ConsoleApp1", "ConsoleApp1.csproj");
         var project1 = new TestPackageSpecFactory(project1Path, builder =>
         {
             builder.WithProperty("TargetFramework", "net9.0")
                    .WithItem("PackageReference", "Test.Package", [new("Version", "1.0.0")])
-                   .WithItem("ProjectReference", @"..\Project2\Project2.csproj", []);
+                   .WithItem("ProjectReference", @"..\ClassLib1\ClassLib1.csproj", []);
         }).Build();
 
-        var project2Path = Path.Combine(solutionDirectory, "Project2", "Project2.csproj");
+        var project2Path = Path.Combine(solutionDirectory, "ClassLib1", "ClassLib1.csproj");
         var project2 = new TestPackageSpecFactory(project2Path, builder =>
         {
             builder.WithProperty("TargetFramework", "net9.0");

